File: database/database-annotation-processor/src/main/java/ru/tinkoff/kora/database/annotation/processor/QueryWithParameters.java
package ru.tinkoff.kora.database.annotation.processor;

import ru.tinkoff.kora.annotation.processor.common.ProcessingErrorException;

import javax.annotation.Nullable;
import javax.annotation.processing.Filer;
import javax.tools.StandardLocation;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.Optional;

public record QueryWithParameters(String rawQuery, List<QueryParameter> parameters) {

    public record QueryParameter(String sqlParameterName, int methodIndex, List<Integer> sqlIndexes) {}

    @Nullable
    public QueryParameter find(String name) {
        for (var parameter : parameters) {
            if (parameter.sqlParameterName.equals(name)) {
                return parameter;
            }
        }
        return null;
    }

    @Nullable
    public QueryParameter find(int methodIndex) {
        for (var parameter : parameters) {
            if (parameter.methodIndex == methodIndex) {
                return parameter;
            }
        }
        return null;
    }

    public static QueryWithParameters parse(Filer filer, String rawSql, List<ru.tinkoff.kora.database.annotation.processor.model.QueryParameter> parameters) {
        if (rawSql.startsWith("classpath:/")) {
            var path = rawSql.substring(11);
            var i = path.lastIndexOf("/");
            final String packageName;
            final String resourceName;
            if (i > 0) {
                packageName = path.substring(0, i).replace('/', '.');
                resourceName = path.substring(i + 1);
            } else {
                packageName = "";
                resourceName = path;
            }
            try (var is = filer.getResource(StandardLocation.SOURCE_PATH, packageName, resourceName).openInputStream()){
                rawSql = new String(is.readAllBytes(), StandardCharsets.UTF_8);
            } catch (IOException e) {
                try (var is = filer.getResource(StandardLocation.CLASS_PATH, packageName, resourceName).openInputStream()){
                    rawSql = new String(is.readAllBytes(), StandardCharsets.UTF_8);
                } catch (IOException e1) {
                    e.addSuppressed(e1);
                    throw new RuntimeException(e);
                }
            }
        }

        List<QueryParameter> params = new ArrayList<>();

        for (int i = 0; i < parameters.size(); i++) {
            var parameter = parameters.get(i);
            var parameterName = parameter.name();

            if (parameter instanceof ru.tinkoff.kora.database.annotation.processor.model.QueryParameter.ConnectionParameter) {
                continue;
            }
            var size = params.size();
            if (parameter instanceof ru.tinkoff.kora.database.annotation.processor.model.QueryParameter.BatchParameter batchParameter) {
                parameter = batchParameter.parameter();
            }
            if (parameter instanceof ru.tinkoff.kora.database.annotation.processor.model.QueryParameter.SimpleParameter simpleParameter) {
                parseSimpleParameter(rawSql, i, parameterName).ifPresent(params::add);
            }
            if (parameter instanceof ru.tinkoff.kora.database.annotation.processor.model.QueryParameter.EntityParameter entityParameter) {
                for (var field : entityParameter.entity().entityFields()) {
                    parseSimpleParameter(rawSql, i, parameterName + "." + field.element().getSimpleName()).ifPresent(params::add);
                }
            }
            if (params.size() == size) {
                throw new ProcessingErrorException("Parameter usage was not found in query: " + parameterName, parameter.variable());
            }
        }

        var paramsNumbers = params
            .stream()
            .map(QueryParameter::sqlIndexes)
            .flatMap(Collection::stream)
            .sorted()
            .toList();

        params = params.stream()
            .map(p -> new QueryParameter(p.sqlParameterName(), p.methodIndex(), p.sqlIndexes()
                .stream()
                .map(paramsNumbers::indexOf)
                .toList()
            ))
            .toList();

        return new QueryWithParameters(rawSql, params);
    }


    private static Optional<QueryParameter> parseSimpleParameter(String rawSql, int methodParameterNumber, String sqlParameterName) {
        int index = -1;
        var result = new ArrayList<Integer>();
        while ((index = rawSql.indexOf(":" + sqlParameterName, index + 1)) >= 0) {
            var indexAfter = index + sqlParameterName.length() + 1;
            if (rawSql.length() >= indexAfter + 1) {
                var charAfter = rawSql.charAt(indexAfter);
                if (Character.isAlphabetic(charAfter) || charAfter == '_' || charAfter == '$' || Character.isDigit(charAfter)) {
                    continue;
                }
            }
            result.add(index);
        }

        return (result.isEmpty())
            ? Optional.empty()
            : Optional.of(new QueryParameter(sqlParameterName, methodParameterNumber, result));
    }
}

File: database/database-symbol-processor/src/main/kotlin/ru/tinkoff/kora/database/symbol/processor/QueryWithParameters.kt
package ru.tinkoff.kora.database.symbol.processor

import ru.tinkoff.kora.ksp.common.exception.ProcessingErrorException
import java.io.BufferedInputStream
import java.nio.charset.Charset

data class QueryWithParameters(val rawQuery: String, val parameters: List<QueryParameter>) {
    data class QueryParameter(val sqlParameterName: String, val methodIndex: Int, val sqlIndexes: List<Int>)

    fun find(name: String): QueryParameter? {
        for (parameter in parameters) {
            if (parameter.sqlParameterName == name) {
                return parameter
            }
        }
        return null
    }

    fun find(methodIndex: Int): QueryParameter? {
        for (parameter in parameters) {
            if (parameter.methodIndex == methodIndex) {
                return parameter
            }
        }
        return null
    }

    companion object {

        fun parse(rq: String, parameters: List<ru.tinkoff.kora.database.symbol.processor.model.QueryParameter>): QueryWithParameters {
            val params = mutableListOf<QueryParameter>()
            var rawSql = rq
            if (rawSql.startsWith("classpath:/")) {
                val file = ClassLoader.getSystemClassLoader().getResource(rawSql.replaceFirst("classpath:/", ""))
                val content = file.content as BufferedInputStream
                rawSql = content.use {
                    it.readAllBytes().toString(Charset.defaultCharset())
                }
            }
            parameters.forEachIndexed { i, _parameter ->
                var parameter = _parameter
                val parameterName = parameter.name
                if (parameter is ru.tinkoff.kora.database.symbol.processor.model.QueryParameter.ConnectionParameter) {
                    return@forEachIndexed
                }
                val size = params.size
                if (parameter is ru.tinkoff.kora.database.symbol.processor.model.QueryParameter.BatchParameter) {
                    parameter = parameter.parameter
                }
                if (parameter is ru.tinkoff.kora.database.symbol.processor.model.QueryParameter.SimpleParameter) {
                    parseSimpleParameter(rawSql, i, parameterName).let {
                        if (it.sqlIndexes.isNotEmpty()) {
                            params.add(it)
                        }
                    }
                }
                if (parameter is ru.tinkoff.kora.database.symbol.processor.model.QueryParameter.EntityParameter) {
                    for (field in parameter.entity.fields) {
                        parseSimpleParameter(rawSql, i, parameterName + "." + field.property.simpleName.getShortName()).let {
                            if (it.sqlIndexes.isNotEmpty()) {
                                params.add(it)
                            }
                        }
                    }
                }
                if (params.size == size) {
                    throw ProcessingErrorException("Parameter usage was not found in sql: ${parameter.name}", parameter.variable)
                }
            }
            val paramsNumbers = params.asSequence()
                .map { it.sqlIndexes }
                .flatten()
                .sorted()
            val processedParams = params
                .map { p: QueryParameter ->
                    QueryParameter(
                        p.sqlParameterName, p.methodIndex, p.sqlIndexes
                            .map { paramsNumbers.indexOf(it) }
                    )
                }

            return QueryWithParameters(rawSql, processedParams)
        }

        private fun parseSimpleParameter(rawSql: String, methodParameterNumber: Int, sqlParameterName: String): QueryParameter {
            var index = -1
            val result = ArrayList<Int>()
            while (rawSql.indexOf(":$sqlParameterName", index + 1).also { index = it } >= 0) {
                val indexAfter = index + sqlParameterName.length + 1
                if (rawSql.length >= indexAfter + 1) {
                    val charAfter = rawSql[indexAfter]
                    if (Character.isAlphabetic(charAfter.code) || charAfter == '_' || charAfter == '$' || Character.isDigit(charAfter)) {
                        continue
                    }
                }
                result.add(index)
            }
            return QueryParameter(sqlParameterName, methodParameterNumber, result)
        }
    }
}
